source/common/common/radix_tree.h

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include "source/common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+template <class Value> class RadixTree {
+  static constexpr int32_t NoNode = -1;
+  struct RadixTreeNode {
+    std::string prefix_;
+    Value value_{};
+    // Hash map for O(1) child lookup by first character
+    absl::flat_hash_map<uint8_t, RadixTreeNode> children_;
+
+    /**
+     * Insert a key-value pair into this node
+     * @param key the full key being inserted
+     * @param search the remaining search key
+     * @param value the value to insert
+     */
+    void insert(absl::string_view key, absl::string_view search, Value value) {
+      // Handle key exhaustion
+      if (search.empty()) {
+        value_ = std::move(value);
+        return;
+      }
+
+      // Look for the edge
+      uint8_t firstChar = static_cast<uint8_t>(search[0]);
+      auto childIt = children_.find(firstChar);
+
+      // No edge, create one
+      if (childIt == children_.end()) {
+        // Create a new child node
+        RadixTreeNode newChild;
+        newChild.prefix_ = std::string(search);
+        newChild.value_ = std::move(value);
+
+        // Add the child to the current node
+        children_[firstChar] = std::move(newChild);
+        return;
+      }
+
+      // Get the child node
+      RadixTreeNode& child = childIt->second;
+
+      // Determine longest prefix of the search key on match
+      size_t commonPrefix = longestPrefix(search, child.prefix_);
```

Comment:
Nit: could this variable and function have 'Length' appended to the names? It currently reads a bit weirdly because e.g. "foot" and "foobar" the longest common prefix is "foo" but this function and value are not "foo", they're 3.

(Maybe `commonPrefixLength` rather than `longestPrefix`, and then name the local variable something short like `cpl` since it's only used within a few lines.)